netket/models/autoreg.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Any, Callable, Iterable, Tuple, Union
+
+from flax import linen as nn
+from jax import numpy as jnp
+from netket.hilbert import CustomHilbert
+from netket.nn import MaskedConv1D, MaskedDense1D
+from netket.nn.initializers import lecun_normal, zeros
+from netket.utils.types import Array, DType, NNInitFunc, PyTree
+
+
+class ARNN(nn.Module):
+    """Base class for autoregressive neural networks."""
+
+    @abc.abstractmethod
+    def conditionals(self, inputs: Array, cache: PyTree) -> Tuple[Array, PyTree]:
+        """
+        Computes the probabilities for each spin to take each value.
+
+        Args:
+          inputs: configurations with dimensions (batch, Hilbert.size).
+          cache: auxiliary states, e.g., used to implement fast autoregressive sampling.
+
+        Returns:
+          p: the probabilities with dimensions (batch, Hilbert.size, Hilbert.local_size).
+          cache: the updated cache.
+        """
+        raise NotImplementedError
+
+
+class ARNNDense(ARNN):
+    """Autoregressive neural network with dense layers."""
+
+    hilbert: CustomHilbert
+    """the discrete Hilbert space."""
+    layers: int
+    """number of layers."""
+    features: Union[Iterable[int], int]
+    """number of features in each layer. If a single number is given,
+    all layers except the last one will have the same number of features."""
+    activation: Callable[[Array], Array] = nn.silu
+    """the nonlinear activation function between hidden layers (default: silu)."""
+    use_bias: bool = True
+    """whether to add a bias to the output (default: True)."""
+    dtype: DType = jnp.float64
```

Comment:
does this network work with complex dtypes?

Reply:
To make complex dtypes work, I need to change the default `kernel_init` to `netket.nn.initializers.lecun_complex`.

Also I would like to change the default `activation` to `jax.nn.selu`, which is consistent with `GCNN`.

Reply:
are you sure this is enough? don't you also need to take care of the fact that you need to normalize according to |marginal|^2 ? I think in the present version you normalize in L1 norm and then take the square root? maybe I am wrong...

Reply:
Previously I usually use SiLU as the default activation function. In NetKet I found `netket.nn.activation.swish`. Could you explain why you need `add_complex_wrap` there? It seems `nkjax.grad(nknn.swish)` behaves the same as `nkjax.grad(jax.nn.swish)`. Also, there is no `add_complex_wrap` for `silu`.

Reply:
Some side notes: It seems 'SiLU' should be the formal name, rather than 'Swish', to give appropriate credit to its inventors. JAX, TensorFlow and PyTorch all choose 'SiLU' as the formal name, and 'Swish' as the alias.

Reply:
@wdphy16 `add_complex_wrap` is there becaues some jax's activation functions don't work with complex numbers. Eg `jax.nn.sigmoid` uses `jnp.log1p` for improved numerical stability which does not support complex numbers. 

I simply added a trivial implementation (not as stable as `log1p`) that works with complex numbers. If you have dieas on how to generalize it to complex numbers with tricks like `log1p`, then we should upstream them to jax and eventually remove our wrap.

It is also possible that recent versions of jax no longer need some of those wrappings.
I did not implement all wrappings (so indeed some are missing).

A PR to add the missing ones is welcome.

Reply:
re initializer: you can write a simple initializer that checks if dtype is real or complex.
If real uses lecun_normal if complex uses lecun_complex (which is a very bad name. It's just a rescaled normal)

Reply:
> are you sure this is enough? don't you also need to take care of the fact that you need to normalize according to |marginal|^2 ? I think in the present version you normalize in L1 norm and then take the square root? maybe I am wrong...

You're right. We also need to change the sigmoid in the final layer to something else.

Reply:
for the initializer, pick what you think is best. you can also make the default depend on whever it's complex or float.

Reply:
> we should upstream them to jax and eventually remove our wrap.

I agree. @PhilipVinc I just found that you're making a PR to JAX about complex `jax.scipy.special.logsumexp`. After finishing that, we can do the same for `jax.numpy.logaddexp`, and `jax.nn.softplus` should work then.

As for `jax.nn.sigmoid`, it calls `jax.scipy.special.expit`, which seems to be already working in the recent version of JAX.

Reply:
Let me write another comment about complex outputs of ARNN...

Reply:
For the initializer, if we want it to be truncated, maybe we can use
```
x = randn(shape) + randn(shape) * 1j
return where(abs(x) < threshold, x, x / abs(x) * threshold)
```
which is well-defined for both real and complex. (Of course it's not a big problem)